ddl/delete_range.go

```diff
@@ -0,0 +1,335 @@
+					if !finish {
+						dr.keys = append(dr.keys, iter.Key().Clone())
+						if err := iter.Next(); err != nil {
+							return errors.Trace(err)
+						}
+						continue
+					}
+				}
+				break
+			}
+			for _, key := range dr.keys {
+				err := txn.Delete(key)
+				if err != nil && !terror.ErrorEqual(err, kv.ErrNotExist) {
+					return errors.Trace(err)
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if finish {
+			break
+		}
+	}
+	return nil
+}
+
+// insertBgJobIntoDeleteRangeTable parses the job into delete-range arguments,
+// and inserts a new record into gc_delete_range table. The primary key is
+// job ID, so we ignore key conflict error.
+func insertBgJobIntoDeleteRangeTable(ctx context.Context, job *model.Job) error {
+	s := ctx.(sqlexec.SQLExecutor)
+	switch job.Type {
+	case model.ActionDropSchema:
+		var tableIDs []int64
+		if err := job.DecodeArgs(&tableIDs); err != nil {
+			return errors.Trace(err)
+		}
+		for _, tableID := range tableIDs {
+			startKey := tablecodec.EncodeTablePrefix(tableID)
+			endKey := tablecodec.EncodeTablePrefix(tableID + 1)
+			if err := doInsert(s, job.ID, tableID, startKey, endKey, time.Now().Unix()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	case model.ActionDropTable, model.ActionTruncateTable:
+		tableID := job.TableID
+		startKey := tablecodec.EncodeTablePrefix(tableID)
+		endKey := tablecodec.EncodeTablePrefix(tableID + 1)
+		return doInsert(s, job.ID, tableID, startKey, endKey, time.Now().Unix())
+	case model.ActionDropIndex:
+		tableID := job.TableID
+		var indexName interface{}
+		var indexID int64
+		if err := job.DecodeArgs(&indexName, &indexID); err != nil {
+			return errors.Trace(err)
+		}
+		startKey := tablecodec.EncodeTableIndexPrefix(tableID, indexID)
+		endKey := tablecodec.EncodeTableIndexPrefix(tableID, indexID+1)
+		return doInsert(s, job.ID, indexID, startKey, endKey, time.Now().Unix())
+	}
+	return nil
+}
+
+func doInsert(s sqlexec.SQLExecutor, jobID int64, elementID int64, startKey, endKey kv.Key, ts int64) error {
+	log.Infof("[ddl] insert into delete-range table with key: (%d,%d)", jobID, elementID)
+	startKeyEncoded := hex.EncodeToString(startKey)
+	endKeyEncoded := hex.EncodeToString(endKey)
+	sql := fmt.Sprintf(insertDeleteRangeSQL, jobID, elementID, startKeyEncoded, endKeyEncoded, ts)
+	_, err := s.Execute(sql)
+	return errors.Trace(err)
+}
+
+// DelRangeTask is for run delete-range command in gc_worker.
+type DelRangeTask struct {
+	jobID, elementID int64
+	startKey, endKey []byte
+}
+
+// Range returns the range [start, end) to delete.
+func (t DelRangeTask) Range() ([]byte, []byte) {
+	return t.startKey, t.endKey
+}
+
+// LoadDeleteRanges loads delete range tasks from gc_delete_range table.
+func LoadDeleteRanges(ctx context.Context, safePoint uint64) (ranges []DelRangeTask, _ error) {
+	sql := fmt.Sprintf(loadDeleteRangeSQL, safePoint)
+	rss, err := ctx.(sqlexec.SQLExecutor).Execute(sql)
```

Comment:
I think function `ExecRestrictedSQL` of `session` is not suite for this case, because in that function, we call `sysSessionPool.Get()` to get a `Session`; if the pool is empty, so we call `createSession` to create one, which will requira lock in `domap`. But at that time, we probably have already in a locked context, so it will be a dead lock.